onnx/onnx-ml.proto

```diff
@@ -188,6 +194,115 @@ message NodeProto {
+  repeated TensorProto additional_initializer = 1;
+
+  // Training-specific inputs, for example, labels in classification problems.
+  // The values can only be inputs of "loss" and "optimizer" defined below.
+  // The full input list of the training computation graph is the concatenation
+  // of "ModelProto.graph.input" and
+  // "ModelProto.training_info.additional_input".
+  repeated ValueInfoProto additional_input = 2;
+
+  // Training-specific outputs, for example, new tensor values computed by the
+  // specified training algorithm. The full output list of the training
+  // computation graph is the concatenation of "ModelProto.graph.output" and
+  // "ModelProto.training_info.additional_output".
+  repeated ValueInfoProto additional_output = 3;
+  
+  // Loss node which consumes tensors in "ModelProto.graph", "additional_inputs"
+  // , and "additional_initializer". Loss nodes produces a scalar loss value.
+  // In training phase, the optimization algorithm would be launch to minimizes
+  // loss node's output value. "loss" can be an ONNX primitive operator or a
+  // function defined in "ModelProto.function."
+  required NodeProto loss = 4;
+  
+  // Optimization algorithm used to minimize the loss function. It can be an
+  // ONNX primitive operator or a function defined in "ModelProto.function."
+  // This node can reference tensors in the "ModelProto.graph",
+  // "additional_inputs," "additional_initializer," and gradient tensor names
+  // defined below.
+  // To connect the optimizer's parameter, optimized tensors, optimized tensors'
+  // gradient, and optimizer correctly without breaking
+  // single-static-assignment, "optimizer" has some restruction on its input
+  // and output schema.
+  //
+  // Restructions on schema:
+  //   1. Input list should be [gradient tensors, optimized tensors, 
+  //      optimizer states, other variables used in the optimizer]. For
+  //      example, to minimize the loss function w.r.t. to tensors X1 and X2,
+  //      the expected input list would be ["grad_X1", "grad_X2", "X1", "X2",
+  //      "accumulated_Grad_X1", "accumulated_Grad_X2", "learning_rate", ...]
+  //   2. The output list's format is [optimized tensors' new values, new
+  //      optimizer states]. Consider the "X1"-"X2" example above again. The
+  //      corresponding output list would be ["new_X1", "new_X2",
+  //      "new_accumulate_grad_X1", "new_accumulate_grad_X2"].
+  //   3. Optimized tensors and optimizer states must be defined in
+  //      "ModelProto.graph.initializer" or
+  //      "ModelProto.training_info.additional_initializer."
+  required NodeProto optimizer = 5;
+
+  // Optimized tensors and their gradient tensor names. Each pair describes a
+  // tensor's name (key) and its gradient's name (value). The gradient tensors
+  // are only visible to the input list of "optimizer."
+  repeated StringStringEntryProto gradient_bindings = 6;
```

Comment:
Just a minor suggestion: can we swap the order of the optimizer and gradient_bindings? Just to correspond to the order of evaluation (cost function first, then computation of gradients, and then the optimizer step).

Reply:
Sure.